src/Kis/Kis.hs

```diff
@@ -2,39 +2,69 @@
 {-# LANGUAGE OverloadedStrings          #-}
 {-# LANGUAGE RankNTypes                 #-}
 {-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+{-# LANGUAGE FlexibleInstances          #-}
 
 module Kis.Kis
     ( withKis
-    , KisAction(..)
+    , KisRequest(..)
     , KisClient
     , KisConfig(..)
     , Kis(..)
     , KisException(..)
+    , KisT
+    , createBed
+    , createPatient
+    , getPatients
+    , getPatient
+    , placePatient
+    , runKis
     )
 where
 
 import Control.Monad.Logger
+import Control.Monad.Operational
 import Control.Monad.RWS hiding (asks)
 import Control.Monad.Reader
 import Database.Persist
-import Data.Text
+import Data.Text hiding (singleton)
 import GHC.Exception
 
 import Kis.Model
 import Kis.Time
 
-data KisAction a where
-    CreateBed :: String -> KisAction BedId
-    CreatePatient :: Patient -> KisAction PatientId
-    GetPatient :: PatientId -> KisAction (Maybe Patient)
-    GetPatients :: KisAction [Entity Patient]
-    PlacePatient :: PatientId -> BedId -> KisAction (Maybe PatientBedId)
+data KisRequest a where
+    CreateBed :: String -> KisRequest BedId
+    CreatePatient :: Patient -> KisRequest PatientId
+    GetPatient :: PatientId -> KisRequest (Maybe Patient)
+    GetPatients :: KisRequest [Entity Patient]
+    PlacePatient :: PatientId -> BedId -> KisRequest (Maybe PatientBedId)
 
-deriving instance Show (KisAction a)
+type KisT m = ProgramT KisRequest m
+
+runKis :: KisT IO a -> IO a
+runKis = undefined
+
+class Monad m => MonadKis m where
+    createBed :: String -> m BedId
+    createPatient :: Patient -> m PatientId
+    getPatients :: m [Entity Patient]
+    getPatient :: PatientId -> m (Maybe Patient)
+    placePatient :: PatientId -> BedId -> m (Maybe PatientBedId)
+
+instance Monad m => MonadKis (KisT m) where
+    createPatient = singleton . CreatePatient
+    createBed = singleton . CreateBed
+    getPatients = singleton GetPatients
+    getPatient = singleton . GetPatient
+    placePatient patId bedId = singleton (PlacePatient patId bedId)
+
+
+deriving instance Show (KisRequest a)
 
```

Comment:
Vielleicht kannst du ja nochmal kurz erklären, was für dich die Vorteile sind? Glaubst du man könnte Read / Write Funktionalität gut trennen, indem man verschiedene Typklassen einführt und dann halt eingeschränkt ist welche Funktionen man benutzen kann? Ich denke, bei diesem Ansatz sollte `Clock` dann auch als Typklasse abstrahiert werden. Ich weiss nicht genau, wie sich diese Modularisierung mit den Sachen aus operational verträgt.

Ich finde es ist wichtig genau zu sagen, welche Objekte im Code was definieren, also ist die Kis API durch `KisRequest a`, oder durch `Kis m` definiert, oder eben durch die `MonadKis`? Im Moment definiert `Kis m` für mich alle möglichen Operationen in `KisClient m a` und `KisRequest a` definiert das Interface zu der Datenschicht. Die `Clock` Funktionalität ist Teil des Kis, also in `Kis m` hat aber nichts mit der Datenschicht zu tun, sollte also nicht in `KisRequest a` sein. Das ist natürlich gerade nur meine Betrachtungsweise, aber ich denke es lohnt sich diese Dinge klar zu machen und dann auch als Kommentar festzuhalten.